### PR TITLE
needless_range_loop: improve documentation

### LINKED ISSUE #6929: Improve documentation of `needless_range_loop`
> I still believe this lint has grown too large, if we look at the doc page, it says "Checks for looping over the range of 0..len of some collection just to get the values by index", so at least the documentation should be updated to discuss any a..b range.

_Originally posted by @ChrisJefferson in https://github.com/rust-lang/rust-clippy/issues/6075#issuecomment-801810194_

We should improve the documentation of the [`needless_range_loop`](https://rust-lang.github.io/rust-clippy/master/index.html#needless_range_loop) lint to better explain what cases this lint triggers on. Also make more clear, that using iterators _usually_ has better performance (re-classify?).

### PR DESCRIPTION
**Pull request by tylerjw**
fixes #6929

changelog: [`needless_range_loop`]: improve documentation

### CODE REVIEW

**Comment by ChrisJefferson** [comment]
I think there are further improvements to be made here, but this is an improvement over what was there before.

**Comment by tylerjw** [comment]
What would you like it to also do?

**Comment by ChrisJefferson** [comment]
Sorry, I should have been clearer.

There is also [#6930](https://github.com/rust-lang/rust-clippy/issues/6930) , which discusses how this needless_range_loop can change behaviour -- however I think that should be dealt with separately, if for no other reason than it's not clear (to me) if the fix should be making this clear in the docs, or changing the suggestion.

**Comment by tylerjw** [comment]
Thank you for clarifying. I read that other issue and I'm unsure what change to make for it, that is why I only linked the issue about clarifying the docs.

File: clippy_lints/src/loops/needless_range_loop.rs [hunk 1]
```diff
@@ -149,7 +149,7 @@ pub(super) fn check<'tcx>(
                         |diag| {
                             multispan_sugg(
                                 diag,
-                                "consider using an iterator",
+                                "consider using an iterator and enumerate",
```
Comment (Manishearth) [inline clippy_lints/src/loops/needless_range_loop.rs:152]:
```suggestion
                                "consider using an iterator and enumerate()",
```

**Review by Manishearth — APPROVED** [review]

File: clippy_lints/src/loops/needless_range_loop.rs (hunk 1 above)
Comment (tylerjw) [reply clippy_lints/src/loops/needless_range_loop.rs]: Pushed commit with bless for this change.

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
@bors r+

thanks!

**Pull request merged**